synapse/federation/sender/per_destination_queue.py

```diff
@@ -217,17 +217,21 @@ def _transaction_transmission_loop(self):
                 pending_edus = []
 
                 pending_edus.extend(self._get_rr_edus(force_flush=False))
+                pending_edus.extend(device_message_edus[:99])
 
                 # We can only include at most 100 EDUs per transactions
-                pending_edus.extend(self._pop_pending_edus(100 - len(pending_edus)))
+                # But we should keep one slot free for presence
+                to_remove = []
+                for key, val in self._pending_edus_keyed.items():
```

Comment:
I think this would be clearer written as a while loop:

```
while len(pending_edus) < 100:
    _, val = self._pending_edus_keyed.popitem()
    pending_edus.append(val)
```